flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamPhysicalWindowAggregate.scala

```diff
@@ -0,0 +1,137 @@
+    cluster: RelOptCluster,
+    traitSet: RelTraitSet,
+    inputRel: RelNode,
+    val grouping: Array[Int],
+    val aggCalls: Seq[AggregateCall],
+    val windowing: WindowingStrategy,
+    val namedWindowProperties: Seq[PlannerNamedWindowProperty])
+  extends SingleRel(cluster, traitSet, inputRel)
+  with StreamPhysicalRel {
+
+  override def requireWatermark: Boolean = windowing.isRowtime
+
+  override def deriveRowType(): RelDataType = {
+    val groupSet = ImmutableBitSet.of(grouping: _*)
+    val baseType = Aggregate.deriveRowType(
+      cluster.getTypeFactory,
+      getInput.getRowType,
+      false,
+      groupSet,
+      Collections.singletonList(groupSet),
+      aggCalls.asJava)
+    val typeFactory = getCluster.getTypeFactory.asInstanceOf[FlinkTypeFactory]
+    val builder = typeFactory.builder
+    builder.addAll(baseType.getFieldList)
+    namedWindowProperties.foreach { namedProp =>
+      // use types from windowing strategy which keeps the precision and timestamp type
+      // cast the type to not null type, because window properties should never be null
+      val timeType = namedProp.property match {
+        case PlannerWindowStart(_) | PlannerWindowEnd(_) =>
+          LogicalTypeUtils.removeTimeAttributes(windowing.timeAttributeType).copy(false)
+        case PlannerRowtimeAttribute(_) | PlannerProctimeAttribute(_) =>
+          windowing.timeAttributeType.copy(false)
+      }
+      builder.add(namedProp.name, typeFactory.createFieldTypeFromLogicalType(timeType))
+    }
+    builder.build()
+  }
+
+  override def explainTerms(pw: RelWriter): RelWriter = {
+    val inputRowType = getInput.getRowType
+    val inputFieldNames = inputRowType.getFieldNames.asScala.toArray
+    super.explainTerms(pw)
+      .itemIf("groupBy", RelExplainUtil.fieldToString(grouping, inputRowType), grouping.nonEmpty)
+      .item("window", windowing.toSummaryString(inputFieldNames))
+      .item("select", RelExplainUtil.streamWindowAggregationToString(
+        inputRowType,
+        grouping,
+        getRowType,
+        aggCalls,
+        namedWindowProperties))
+  }
+
+  override def copy(
+      traitSet: RelTraitSet,
+      inputs: util.List[RelNode]): RelNode = {
+    new StreamPhysicalWindowAggregate(
+      cluster,
+      traitSet,
+      inputs.get(0),
+      grouping,
+      aggCalls,
+      windowing,
+      namedWindowProperties
+    )
+  }
+
+  override def translateToExecNode(): ExecNode[_] = {
+    val conf = FlinkRelOptUtil.getTableConfigFromContext(this).getConfiguration
+    if (conf.getBoolean(TABLE_EXEC_EMIT_EARLY_FIRE_ENABLED) ||
+      conf.getBoolean(TABLE_EXEC_EMIT_LATE_FIRE_ENABLED)) {
+      throw new TableException("Currently, window table function based aggregate doesn't " +
+        s"support early-fire and late-fire configuration " +
+        s"'${TABLE_EXEC_EMIT_EARLY_FIRE_ENABLED.key()}' and " +
+        s"'${TABLE_EXEC_EMIT_LATE_FIRE_ENABLED.key()}'.")
+    }
+    new StreamExecWindowAggregate(
```

Comment:
It's better to add check for allow lateness it seems not to be supported yet.

Reply:
We can support this in another PR, this can't be supported in little code. I don't want to make this PR larger.

Reply:
What do you mean about "add check for allow lateness "? From my understanding, we have checked the `TABLE_EXEC_EMIT_LATE_FIRE_ENABLED`.

Reply:
`AllowLateness` now only depends on `tableConfig.getMinIdleStateRetentionTime` instead of `TABLE_EXEC_EMIT_LATE_FIRE_ENABLED ` currently, please see `WindowEmitStrategy`.

Reply:
Yes. This is a bug and should be fixed in another PR.